src/detector/yin.rs

```diff
@@ -0,0 +1,80 @@
+use crate::detector::internals::pitch_from_peaks;
+use crate::detector::internals::DetectorInternals;
+use crate::detector::internals::Pitch;
+use crate::detector::PitchDetector;
+use crate::float::Float;
+use crate::utils::buffer::square_sum;
+use crate::utils::peak::PeakCorrection;
+
+use super::internals::{windowed_square_error, yin_normalize_square_error};
+
+/// Pitch detection based on the YIN algorithm. See http://recherche.ircam.fr/equipes/pcm/cheveign/ps/2002_JASA_YIN_proof.pdf
+pub struct YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    internals: DetectorInternals<T>,
+}
+
+impl<T> YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    pub fn new(size: usize, padding: usize) -> Self {
+        let internals = DetectorInternals::new(1, 3, size, padding);
+        YINDetector { internals }
+    }
+}
+
+impl<T> PitchDetector<T> for YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    fn get_pitch(
+        &mut self,
+        signal: &[T],
+        sample_rate: usize,
+        power_threshold: T,
+        clarity_threshold: T,
+    ) -> Option<Pitch<T>> {
+        // The YIN paper uses 0.1 as a threshold; TarsosDSP uses 0.2. `threshold` is not quite
```

Comment:
I wonder if as more detectors are implemented, we should rethink the signature of `get_pitch` so that different detectors can have a different number/type of parameters.

Not to be fixed in this PR, but maybe something to think about long term

Reply:
I agree. Also, the more sophisticated algorithms keep track of the pitch and use the old pitch to determine the next one. I'm not sure if we want to give `detect_pitch` a memory, or pass in an array of previous pitches if you want to use that for the next guess.